curp/src/server/bg_tasks.rs

```diff
@@ -125,6 +118,73 @@ pub(super) async fn run_bg_tasks<
     info!("all background task stopped");
 }
 
+/// Do periodical task
+#[allow(clippy::integer_arithmetic)] // can't overflow
+async fn bg_tick<C: Command + 'static, Conn: ConnectInterface, ExeTx: CmdExeSenderInterface<C>>(
+    connects: Vec<Arc<Conn>>,
+    state: StateRef<C, ExeTx>,
+    timeout: Arc<ServerTimeout>,
+) {
+    let (rpc_timeout, heartbeat_interval, follower_timeout_ticks, candidate_timeout_ticks) = (
+        *timeout.rpc_timeout(),
+        *timeout.heartbeat_interval(),
+        *timeout.follower_timeout_ticks(),
+        *timeout.candidate_timeout_ticks(),
+    );
+    // randomize to minimize vote split possibility
+    let follower_timeout_ticks =
+        follower_timeout_ticks + thread_rng().gen_range(0..follower_timeout_ticks);
+    let candidate_timeout_ticks =
+        candidate_timeout_ticks + thread_rng().gen_range(0..follower_timeout_ticks);
+
+    // wait for some random time before tick starts to minimize vote split possibility
+    let rand = thread_rng()
+        .gen_range(0..heartbeat_interval.as_millis())
+        .numeric_cast();
+    tokio::time::sleep(Duration::from_millis(rand)).await;
+
+    let mut ticker = tokio::time::interval(heartbeat_interval);
+    ticker.set_missed_tick_behavior(MissedTickBehavior::Skip);
+    loop {
+        let _now = ticker.tick().await;
+        let task = {
+            let state_c = Arc::clone(&state);
+            let state_r = state.upgradable_read();
+            if state_r.is_leader() {
+                if state_r
+                    .hb_opt
+                    .compare_exchange(true, false, Ordering::Acquire, Ordering::Relaxed)
+                    .is_err()
+                {
+                    let resps = bcast_heartbeats(connects.clone(), state_r, rpc_timeout);
+                    Either::Left(handle_heartbeat_responses(
+                        resps,
+                        state_c,
+                        Arc::clone(&timeout),
+                    ))
+                } else {
+                    continue;
+                }
+            } else {
+                let tick = state_r.election_tick.fetch_add(1, Ordering::SeqCst);
+                let timeout_ticks = if state_r.role() == ServerRole::Follower {
+                    follower_timeout_ticks
+                } else {
+                    candidate_timeout_ticks
+                };
+                if tick < timeout_ticks {
+                    continue;
+                }
+                state_r.election_tick.store(0, Ordering::Relaxed); // reset tick
```

Comment:
Why don't you call `state_r.reset_election_tick` here?

Reply:
Fixed